plutus-core/plutus-core/test/Evaluation/Spec.hs

```diff
@@ -51,24 +42,76 @@ test_builtinsDon'tThrow =
+    gen bn = Gen.choice [genArgsWellTyped bn, genArgsArbitrary bn]
+    f bn args = \case
+        Left e -> do
+            annotate "Builtin function evaluation failed"
+            annotate $ "Function: " <> display bn
+            annotate $ "Arguments: " <> display args
+            annotate $ "Error " <> show e
+            failure
+        Right _ -> success
+
+data AlwaysThrows
+    = -- | A builtin function whose denotation always throws an exception.
+      AlwaysThrows
+    deriving stock (Eq, Ord, Show, Bounded, Enum, Ix)
+
+instance Pretty AlwaysThrows where
+    pretty = pretty . show
+
+instance uni ~ DefaultUni => ToBuiltinMeaning uni AlwaysThrows where
+    type CostingPart uni AlwaysThrows = ()
+    toBuiltinMeaning AlwaysThrows = makeBuiltinMeaning f (\_ _ -> mempty)
+      where
+        f :: Integer -> Integer
+        f _ = error "This builtin function always throws an exception."
+
+{- | This test verifies that if evaluating a builtin function actually throws an exception,
+ we'd get a `Left` value, which would cause `test_builtinsDon'tThrow` to fail.
+-}
+test_alwaysThrows :: TestTree
+test_alwaysThrows =
+    testGroup
+        "Builtins throwing exceptions should cause tests to fail"
+        [ testProperty (display AlwaysThrows) $
+            prop_builtinEvaluation @AlwaysThrows AlwaysThrows genArgsWellTyped f
+        ]
+  where
+    f bn args = \case
+        Left _ -> success
+        Right _ -> do
+            annotate "Except builtin function evaluation to throw exceptions, but it didn't"
+            annotate $ "Function: " <> display bn
+            annotate $ "Arguments: " <> display args
+            failure
+
+prop_builtinEvaluation ::
     forall fun.
     (ToBuiltinMeaning DefaultUni fun, Pretty fun) =>
     fun ->
+    -- | A function making a generator for @fun@'s arguments.
+    (fun -> Gen [Term fun]) ->
+    -- | A function that takes a builtin function, a list of arguments, and the evaluation
+    -- outcome, and decides whether to pass or fail the property.
+    (fun -> [Term fun] -> Either SomeException (MakeKnownM (Term fun)) -> PropertyT IO ()) ->
     Property
-prop_builtinsDon'tThrow bn = property $ do
-    args <- forAllNoShow . Gen.choice $ [genArgsWellTyped bn, genArgsArbitrary bn]
-    mbErr <-
-        liftIO $
-            catch
-                (($> Nothing) . evaluate . runEmitter . runExceptT $ eval args)
-                (pure . pure)
-    whenJust mbErr $ \(e :: SomeException) -> do
-        annotate "Builtin function evaluation failed"
-        annotate $ "Function: " <> display bn
-        annotate $ "Arguments: " <> display args
-        annotate $ "Error " <> show e
-        failure
+prop_builtinEvaluation bn mkGen f = property $ do
+    args <- forAllNoShow (mkGen bn) -- Gen.choice $ [genArgsWellTyped bn, genArgsArbitrary bn]
+    f bn args
+        =<< liftIO
+            ( catch @SomeException
+                (fmap Right . evaluate $ eval args)
+                (pure . Left)
```

Comment:
`f bn args =<< liftIO (try @SomeException . evaluate $ eval args)`?

I wonder if we should have any policy regarding catching exceptions. `catch`/`try` here would catch a `SIGTERM` and swallow it, which is not how things are supposed to be, but alternatives are also tricky and for tests we don't care too much.

Reply:
Oh right, should use `try` here.

> I wonder if we should have any policy regarding catching exceptions.

My understanding is that pure exceptions simply shouldn't be caught, period. We catch it here simply for the purpose of printing something useful before letting the test fail.

Reply:
> My understanding is that pure exceptions simply shouldn't be caught, period.

Even specific ones or just the general `SomeException`?

> We catch it here simply for the purpose of printing something useful before letting the test fail

But you don't rethrow the exception, so the tests will continue to execute?

Reply:
> Even specific ones or just the general SomeException?

Any exception. There's no pure version of `catch` in Haskell; you have to catch it in IO, which is why I'm calling `evaluate` and then catching the exception, which is, as I understand it, something you are never supposed to do in non-test code.

> But you don't rethrow the exception, so the tests will continue to execute?

I'm calling `failure` after catching the exception which makes the test fail.

Reply:
> Any exception. There's no pure version of `catch` in Haskell; you have to catch it in IO, which is why I'm calling `evaluate` and then catching the exception, which is, as I understand it, something you are never supposed to do in non-test code.

You're never supposed to call `catch`? Should a call `x / y` kill your entire app? Or do you mean you shouldn't catch exceptions in pure code? Well, people do abuse exceptions and it makes sense to catch an exception purely (via `unsafePerformIO`) and wrap it into an explicit `Either` when you're provided with an unreasonably impure API. Sometimes it's just convenient to design things this way as opposed to wrapping someone else's library, `Data.Text` does it [here](https://hackage.haskell.org/package/text-2.0/docs/src/Data.Text.Encoding.html#decodeUtf8%27) for example. Performance-critical code, especially if it's number crunching, may have no option but to use exceptions, because they are faster than using `Either` or any other kind of pure error handling. We had to move from pure error handling to exceptions in the CEK machine for this reason, for example, and the machine looks pure to the outside. Another concern is laziness: you may want to generate a data structure lazily and if your generator is allowed to fail, you won't be able to release elements to the outside in a lazy fashion if your error handling happens in `Either` or something similar.

So I think there's lots of situations where catching errors in pure code (i.e. with `unsafePerformIO` involved) makes perfect sense. And in IO catching errors is pretty much inevitable, a file not being there shouldn't kill your app. I do agree that catching errors in pure code is something that generally should be avoided.

> I'm calling `failure` after catching the exception which makes the test fail.

This one -- yes, but `failure` does not throw an exception, it just signals that the test failed. Other tests continue to run as normal and your SIGTERM is swallowed by the test. If it wasn't caught, SIGTERM would kill the entire executable running the tests.

Reply:
I meant you shouldn't catch exceptions in pure code, but yeah I shouldn't say never - it should be frowned upon in most application code, but there are certainly valid use cases in low level/library/performance critical code, just like `unsafeCoerce`, `unsafeThaw`, or `unsafeAnything`.

> Other tests continue to run as normal and your SIGTERM is swallowed by the test.

Ah yeah, we certainly shouldn't swallow SIGTERM (or any async exception). But I _think_ hedgehog runs tests in non-main threads (even for `checkSequential`) so this shouldn't prevent SIGTERM from working, but either way we should re-throw `AsyncException`s just to be safe.

Reply:
> I meant you shouldn't catch exceptions in pure code, but yeah I shouldn't say never - it should be frowned upon in most application code, but there are certainly valid use cases in low level/library/performance critical code, just like `unsafeCoerce`, `unsafeThaw`, or `unsafeAnything`.

Ok, we're on the same page then.

> Ah yeah, we certainly shouldn't swallow SIGTERM (or any async exception). But I _think_ hedgehog runs tests in non-main threads (even for `checkSequential`) so this shouldn't prevent SIGTERM from working, but either way we should re-throw `AsyncException`s just to be safe.

Good point. But then they've probably arranged propagation of `ThreadKilled` from the main thread to child threads? And then catching it might mess with some exception-based finalization of tests running in child threads or something... Yeah, we probably shouldn't worry too much about that.

Reply:
> In this case since we actually want to test for the presence of the exception I don't think we have any alternative to catching it!

Sure, but we could catch only this particular exception instead of all of `SomeException`. And in the `DefaultFun` case we could avoid catching anything at all: if it fails, it fails, which is want we want and as I just learned from you `annotate` is a no-op if the test passes. Or we could rethrow the caught exception.

But as I said, I don't think we should worry about that stuff here. There's way too many important things that need to be worked on, so who cares about perfect exception handling in tests code.